File: android/mobileclient/app/src/main/java/com/auditpro/mobile_client/pages/LoginPage.java
/*
  AuditPRO Mobile Client Android
  Copyright 2018 AuditPRO All Rights Reserved
 */
package com.auditpro.mobile_client.pages;

import android.content.Context;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.auditpro.mobile_client.MainActivity;
import com.auditpro.mobile_client.api.ApiClient;
import com.auditpro.mobile_client.api.UserResponse;
import com.auditpro.mobile_client.security.Security;
import com.auditpro.mobile_client.test.R;

import java.lang.ref.WeakReference;


/**
 * Implements the login "page" as a fragment.
 * @author Eric Ruck
 */
@SuppressWarnings("ConstantConditions")
public class LoginPage extends BasePage {

	/**
	 * Required empty public constructor.
	 */
	public LoginPage() { }

	/**
	 * Creates a new instance to get a real login from a cached session.
	 */
	public static LoginPage sessionInstance() {
		LoginPage fragment = new LoginPage();
		Bundle args = new Bundle();
		args.putBoolean(ARG_SESSION_FLAG, true);
		fragment.setArguments(args);
		return fragment;
	}

	/**
	 * Handles initial creation.
	 * @param savedInstanceState Optional state
	 */
	@Override
	public void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		if (getArguments() != null) {
			sessionFlag = getArguments().getBoolean(ARG_SESSION_FLAG, false);
		}
	}

	/**
	 * Creates the view for the login page fragment.
	 * @param inflater Layout inflater to use
	 * @param container Parent for inflated view
	 * @param savedInstanceState Optional state
	 * @return Created view
	 */
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
							 Bundle savedInstanceState) {
		// Inflate the layout for this fragment
		return inflater.inflate(R.layout.fragment_login_page, container, false);
	}

	/**
	 * Initialize the view once it has been created.
	 * @param view Created view
	 * @param savedInstanceState Optional state
	 */
	@Override
	public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);

		// Wire up controls
		emailText = getView().findViewById(R.id.emailText);
		passwordText = getView().findViewById(R.id.passwordText);
		savePasswordCheck = getView().findViewById(R.id.savePasswordCheck);
		passwordText.setOnEditorActionListener(new TextView.OnEditorActionListener() {
			@Override
			public boolean onEditorAction(TextView textView, int i, KeyEvent keyEvent) {
				if ((i == EditorInfo.IME_ACTION_UNSPECIFIED) && (keyEvent != null) &&
						(keyEvent.getKeyCode() == KeyEvent.KEYCODE_ENTER)) {
					if (keyEvent.getAction() == KeyEvent.ACTION_DOWN) {
						return true;
					}
					i = EditorInfo.IME_ACTION_DONE;
				}
				if (i == EditorInfo.IME_ACTION_DONE) {
					if (!loginButton.isEnabled()) {
						// Still need to enter a valid e-mail
						emailText.requestFocus();
					} else {
						// Attempt login
						onLogin();
					}
					return true;
				}
				return false;
			}
		});
		passwordText.addTextChangedListener(new TextWatcher() {
			@Override public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) { }
			@Override public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) { }
			@Override public void afterTextChanged(Editable editable) {
				onLoginInputChanged();
			}
		});
		loginButton = getView().findViewById(R.id.loginButton);
		loginButton.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View view) {
				onLogin();
			}
		});
		getView().findViewById(R.id.helpButton).setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View view) {
				onHelp();
			}
		});

		// Can we set the version?
		String version = getParent().getVersion();
		TextView versionView = getView().findViewById(R.id.versionText);
		if (version != null) {
			// Set the version
			versionView.setText(
					getString(R.string.app_version, version)
			);
		} else {
			// Clear the version, warning already logged
			versionView.setText("");
		}
	}

	/**
	 * Cancels login task in progress on stop.
	 */
	@Override
	public void onStop() {
		super.onStop();
		ConnectedLoginTask inProgress = (loginTask != null) ? loginTask.get() : null;
		if (inProgress != null) {
			inProgress.cancel(true);
		}
	}

	/**
	 * Updates the user interface when the page appears.
	 */
	@Override
	public void onPageAppearing() {
		// Calls default implementation
		super.onPageAppearing();
		if (preserveInput) {
			// Don't reset the input
			preserveInput = false;
		} else {
			// Reinitialize our edits
			Security sec = new Security(getContext().getApplicationContext());
			String savedPassword = sec.getSavedPassword();
			emailText.setText(sec.getLastEmail());
			passwordText.setText(savedPassword);
			savePasswordCheck.setChecked(savedPassword.length() > 0);
			onLoginInputChanged();
		}
	}

	/**
	 * Updates the login button depending on the validity of the input state.
	 */
	private void onLoginInputChanged() {
		String email = emailText.getText().toString();
		String password = passwordText.getText().toString();
		boolean isLoginEnabled = (password.length() > 0) &&
				(email.matches(EMAIL_REGEX));
		loginButton.setEnabled(isLoginEnabled);
	}

	/**
	 * Handles request to login using entered credentials
	 */
	private void onLogin() {
		// Make sure the keyboard is dismissed
		InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
		imm.hideSoftInputFromWindow(passwordText.getWindowToken(), 0);

		// Do we have a network connection?
		if (((MainActivity) getActivity()).isNetworkConnected()) {
			loginConnected();
		} else if (sessionFlag) {
			Toast.makeText(getContext(), R.string.message_login_inet_req, Toast.LENGTH_SHORT).show();
		} else {
			loginOffline();
		}
	}

	/**
	 * Provides reference-safe async task to login on a worker thread.
	 */
	private static class ConnectedLoginTask extends AsyncTask<Void, Void, UserResponse> {
		private String email;
		private String password;
		private boolean isPasswordSaved;
		private WeakReference<LoginPage> host;
		private ApiClient api;

		/**
		 * Initialize task to call the remote log in web service.
		 * @param host Parent host page
		 */
		ConnectedLoginTask(LoginPage host) {
			this.email = host.emailText.getText().toString();
			this.password = host.passwordText.getText().toString();
			this.isPasswordSaved = host.savePasswordCheck.isChecked();
			this.host = new WeakReference<>(host);
		}

		/**
		 * Shows activity spinner while we run.
		 */
		@Override
		protected void onPreExecute() {
			LoginPage page = host.get();
			IPageParent parent = (page == null) ? null : page.getParent();
			if (parent != null) {
				parent.setActivity(true);
			}
		}

		/**
		 * Execute remove login web service
		 * @param voids Placeholder for no parameters
		 * @return Web service result
		 */
		@Override
		protected UserResponse doInBackground(Void... voids) {
			api = new ApiClient(null);
			return api.login(email, password);
		}

		/**
		 * Handles the web service result in the user interface.
		 * @param userResponse Web service result
		 */
		@Override
		protected void onPostExecute(UserResponse userResponse) {
			super.onPostExecute(userResponse);
			LoginPage self = host.get();
			IPageParent parent = (self == null) ? null : self.getParent();
			if (parent == null) {
				// The page was popped before we completed
				return;
			}
			parent.setActivity(false);
			if (userResponse == null) {
				// Failed to login, show error toast
				Toast.makeText(self.getContext(), api.getMessage(), Toast.LENGTH_LONG).show();
			} else {
				// Complete login
				((MainActivity) self.getActivity()).setLogin(api.getToken(), email, password, isPasswordSaved, userResponse);
				if (self.sessionFlag) {
					// Return to the calling page
					parent.popPage();
				} else {
					// Go to the main menu
					parent.pushPage(new MainMenuPage());
				}
			}
		}
	}

	private WeakReference<ConnectedLoginTask> loginTask;

	/**
	 * Validates the login against remote web services.
	 */
	private void loginConnected() {
		loginTask = new WeakReference<>(new ConnectedLoginTask(this));
		loginTask.get().execute();
	}

	/**
	 * Logs in the user with cached credentials.
	 */
	private void loginOffline() {
		// Do we have cached credentials?
		Security security = new Security(getContext().getApplicationContext());
		String lastEmail = security.getLastEmail();
		if (lastEmail == null)  {
			// No cached credentials
			Toast.makeText(getContext(), R.string.message_login_inet_first, Toast.LENGTH_LONG).show();
		} else if (!lastEmail.toLowerCase().equals(emailText.getText().toString().toLowerCase()) ||
				!security.verifyLastPassword(passwordText.getText().toString(), savePasswordCheck.isChecked())) {
			// Credentials don't match
			Toast.makeText(getContext(), R.string.message_login_incorrect, Toast.LENGTH_SHORT).show();
		} else {
			// We're logged in
			IPageParent parent = getParent();
			if (parent != null) {
				// Show the main menu
				parent.pushPage(new MainMenuPage());
			}
		}
	}

	/**
	 * Handles request to show help.
	 */
	private void onHelp() {
		IPageParent parent = getParent();
		if (parent != null) {
			// Show the help screen
			preserveInput = true;
			parent.pushPage(HelpPage.newInstance("LoginHelp.html"));
		}
	}

	private static final String ARG_SESSION_FLAG = "sessionFlag";
	private boolean sessionFlag;

	private EditText emailText;
	private EditText passwordText;
	private Button loginButton;
	private CheckBox savePasswordCheck;
	private boolean preserveInput;

	private static final String EMAIL_REGEX = "\\A(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?)\\Z";
}

File: ios/Mobile Client/Mobile Client/pages/LoginPage.swift
//
//  LoginPage.swift
//  Mobile Client
//
//  Created by Eric Ruck on 1/6/18.
//  Copyright 2018-2019 AuditPro. All rights reserved.
//

import UIKit


/**
 * Implements the login page controller.
 * @author Eric Ruck
 */
class LoginPage: BasePage, UITextFieldDelegate {

	// Control outlets
	@IBOutlet weak var emailText: UITextField!
	@IBOutlet weak var passwordText: UITextField!
	@IBOutlet weak var saveSwitch: UISwitch!
	@IBOutlet weak var loginButton: UIButton!
	@IBOutlet weak var versionLabel: UILabel!

	private var sessionFlag: Bool = false
	private var preserveInput: Bool = false
	private var client: ApiClient? = nil


	/**
	 * Initializes the session flag optionally to true, to indicate that we're
	 * showing the login within an offline session, rather that at the start
	 * of the application.
	 * @param sessionFlag In session flag
	 */
	convenience init(sessionFlag: Bool = false) {
		self.init()
		self.sessionFlag = sessionFlag
	}

	/**
	 * Updates when we attach to the parent.
	 * @param parent Parent view controller
	 */
	override func didMove(toParent parent: UIViewController?) {
		// Initialize the version
		versionLabel.text = String.init(format: versionLabel.text!, host!.version)
	}

	/**
	 * Handles tap on the login button.
	 * @param sender Login button
	 */
	@IBAction func onLogin() {
		// Make sure the keyboard is dismissed
		view.endEditing(true)

		// Do we have a network connection?
		if (AppDelegate.isNetworkConnected) {
			loginConnected()
		} else if (sessionFlag) {
			view.makeToast(message: "Please connect to the Internet to log in.")
		} else {
			loginOffline()
		}
	}

	/**
	 * Handles tap on the help button.
	 * @param sender Help button
	 */
	@IBAction func onHelp() {
		preserveInput = true
		host!.pushPage(HelpPage(helpFile: "LoginHelp.html"))
	}

	/**
	 * Updates the enabled state of the login button depending on the validity
	 * of the text field input state.
	 */
	@IBAction func textFieldEditingDidChange(sender: UITextField) {
		let emailTest = NSPredicate(format:"SELF MATCHES[c] %@",
			"(?:[a-z0-9!#$%\\&'*+/=?\\^_`{|}~-]+(?:\\.[a-z0-9!#$%\\&'*+/=?\\^_`{|}"+"~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\"+"x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-"+"z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\[(?:(?:25[0-5"+"]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-"+"9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21"+"-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])+)\\])")
		let isEmailValid =
			emailTest.evaluate(with: emailText.text!)
		let isPasswordValid =
			(passwordText.text!.count > 0)
		NSLog("email isValid: %@, password valid %@", isEmailValid ? "true" : "false", isPasswordValid ? "true":"false")
		loginButton.isEnabled = isEmailValid && isPasswordValid
	}

	/**
	 * Handles completion input on text field.
	 * @param textField Control on which event occurred
	 * @return Perform default flag
	 */
	func textFieldShouldReturn(_ textField: UITextField) -> Bool {
		if (textField == emailText) {
			if (emailText.text!.count > 0) {
				passwordText.becomeFirstResponder()
			}
			return false
		}
		if (textField == passwordText) {
			if (passwordText.text!.count > 0) {
				if (loginButton.isEnabled) {
					onLogin()
				} else {
					emailText.becomeFirstResponder()
				}
			}
			return false
		}

		// Allow default handling
		return true
	}

	/**
	 * Updates the user interface when the page appears.
	 */
	override func onPageAppearing() {
		// Calls default implementation
		super.onPageAppearing()
		if (preserveInput) {
			// Don't reset the input
			preserveInput = false;
		} else {
			// Reinitialize our edits
			let savedPassword = Security.savedPassword
			emailText.text = Security.lastEmail
			passwordText.text = savedPassword
			saveSwitch.isOn = savedPassword.count > 0
			textFieldEditingDidChange(sender: emailText)
		}
	}

	/**
	 * Ensures background tasks are canceled when the page is disappearing.
	 */
	override func onPageDisappearing() {
		client?.cancel()
	}

	/**
	 * Validates the login against remote web services.
	 */
	private func loginConnected() {
		// Show activity
		host!.isActivityShowing = true

		// Async login to web service
		let email = emailText.text!
		let password = passwordText.text!
		client = ApiClient()
		client!.login(email: email, password: password, completion: { (client) in
			// Are we still showing?
			self.client = nil
			if (self.host == nil) {
				return
			}
			DispatchQueue.main.async {
				// Hide activity
				self.host!.isActivityShowing = false

				// Evaluate response
				if (client.message != nil) {
					// Failed to login, show error toast
					self.view.makeToast(message: client.message!)
				} else {
					// Complete login
					AppDelegate.sessionToken = client.token
					Security.setLogin(email: email, password: password,
						isPasswordSaved: self.saveSwitch.isOn,
						userResponse: client.response as! UserResponse)
					if (self.sessionFlag) {
						// Return to the calling page
						self.host!.popPage()
					} else {
						// Go to the main menu
						self.host!.pushPage(MainMenuPage())
					}
				}
			}
		})
	}

	/**
	 * Logs in the user with cached credentials.
	 */
	private func loginOffline() {
		// Do we have cached credentials?

		let lastEmail = Security.lastEmail
		if (lastEmail == "")  {
			// No cached credentials
			view.makeToast(message: "Please connect to the Internet to log in for the first time.")
		} else if ((lastEmail.caseInsensitiveCompare(emailText.text!) == .orderedSame) ||
				!Security.verifyLastPassword(password: passwordText.text!, isPasswordSaved: saveSwitch!.isOn)) {
			// Credentials don't match
			view.makeToast(message: "Your e-mail or password is not correct.")
		} else {
			// We're logged in
			host!.pushPage(MainMenuPage())
		}
	}
}
